share/smack/prelude.py

```diff
@@ -0,0 +1,250 @@
+eq_ops = ['eq', 'ne']
+
+binary_ops_body = ['i1 + i2', 'i1 - i2', 'i1 * i2']
+comp_ops_body = ['i1 <= i2', 'i1 < i2', 'i1 >= i2', 'i1 > i2', 'i1 <= i2', 'i1 < i2', 'i1 >= i2', 'i1 > i2']
+eq_ops_body = ['i1 == i2', 'i1 != i2']
+
+def declare_each_type(func_decl, *args):
+  bpl = ''
+  for i in sizes:
+    tmp = ((i,) + args)
+    bpl += func_decl(*tmp).toStr()
+  return bpl
+
+def declare_each_type_nested(func_decl, *args):
+  bpl = ''
+  for i in range(len(sizes)):
+    for j in range(0,i):
+      tmp = (sizes[i], sizes[j]) + args
+      bpl += func_decl(*tmp).toStr()
+  return bpl
+
+def generate_prelude(args):
+  bpl = ''
+  for i in range(len(binary_ops)):
+    op_name = '$' + binary_ops[i]
+    bpl += declare_each_type(BVBUILTIN_BINARY_OP, 'bv' + binary_ops[i], op_name)
+    if 0 <= i <= 2:
+      bpl += declare_each_type(INLINE_BINARY_OP, op_name, binary_ops_body[i])
+    elif 3 <= i <= 7:
+      bpl += declare_each_type(BUILTIN_BINARY_OP, binary_ops[i][1:], op_name)
+    else:
+      bpl += declare_each_type(UNINTERPRETED_BINARY_OP, op_name)
+
+  for i in range(len(comp_ops)):
+    op_name = '$' + comp_ops[i]
+    bpl += declare_each_type(BVBUILTIN_BINARY_COMP, 'bv' + comp_ops[i], op_name)
+    bpl += declare_each_type(INLINE_BVBUILTIN_BINARY_PRED, op_name)
+    bpl += declare_each_type(INLINE_BINARY_COMP, op_name, comp_ops_body[i])
+    bpl += declare_each_type(INLINE_BINARY_PRED, op_name)
+
+  for i in range(len(eq_ops)):
+    op_name = '$' + eq_ops[i]
+    bpl += declare_each_type(INLINE_BINARY_BV_COMP, op_name, eq_ops_body[i])
+    bpl += declare_each_type(INLINE_BINARY_BV_PRED, op_name)
+    bpl += declare_each_type(INLINE_BINARY_COMP, op_name, eq_ops_body[i])
+    bpl += declare_each_type(INLINE_BINARY_PRED, op_name)
+
+  bpl += declare_each_type(INLINE_BVBUILTIN_BINARY_SELECT, '$min', '$slt')
+  bpl += declare_each_type(INLINE_BVBUILTIN_BINARY_SELECT, '$max', '$sgt')
+  bpl += declare_each_type(INLINE_BVBUILTIN_BINARY_SELECT, '$umin', '$ult')
+  bpl += declare_each_type(INLINE_BVBUILTIN_BINARY_SELECT, '$umax', '$ugt')
+
+  bpl += declare_each_type(INLINE_BINARY_OP, '$smin', '$min(i1,i2)')
+  bpl += declare_each_type(INLINE_BINARY_OP, '$smax', '$max(i1,i2)')
+  bpl += declare_each_type(INLINE_BINARY_OP, '$umin', '$min(i1,i2)')
+  bpl += declare_each_type(INLINE_BINARY_OP, '$umax', '$max(i1,i2)')
+
+  bpl += declare_each_type(BVBUILTIN_UNARY_OP, 'bvnot', '$not')
+  bpl += declare_each_type(UNINTERPRETED_UNARY_OP, '$not')
+
+  bpl += declare_each_type_nested(TRUNC_BV_OP)
+  bpl += declare_each_type_nested(TRUNC_OP)
+
+  bpl += declare_each_type_nested(INLINE_CONVERSION, '$zext', 'i1')
+  bpl += declare_each_type_nested(INLINE_CONVERSION, '$sext', 'i1')
+
+  bpl += declare_each_type(SAFE_LOAD_OP, 'i1[i2]')
+  bpl += declare_each_type(SAFE_LOAD_BV_OP, 'i1[i2]')
+  bpl += declare_each_type(SAFE_STORE_OP, 'i1[i2 := i3]')
+  bpl += declare_each_type(SAFE_STORE_BV_OP, 'i1[i2 := i3]')
+
+  f = open(args.bpl_file, 'a+')
```

Comment:
Perhaps you can separate the writing of a given prelude to a file, e.g., `append_prelude`, from the construction of a prelude, e.g., `get_prelude`.